python/ray/_private/package.py

```diff
@@ -0,0 +1,160 @@
+    # maintain the lifetime of this object
+    PIN_GCS = "pingcs"
+
+def _xor_bytes(left, right):
+    if left and right:
+        return bytes(a ^ b for (a, b) in zip(left, right))
+    return left or right
+
+def _zip_module(path, relative_path, zip_handler):
+    for from_file_name in path.glob("**/*"):
+        to_file_name = from_file_name.relative_to(relative_path)
+        zip_handler.write(from_file_name, to_file_name)
+
+def _hash_modules(path):
+    hash_val = None
+    BUF_SIZE = 4096 * 1024
+    for from_file_name in path.glob("**/*"):
+        md5 = hashlib.md5()
+        if Path(from_file_name).is_dir():
+            md5.update(str(from_file_name).encode())
+        else:
+            with open(from_file_name, mode="rb") as f:
+                data = f.read(BUF_SIZE)
+                if not data:
+                    break
+                md5.update(data)
+        hash_val = _xor_bytes(hash_val, md5.digest())
+    return hash_val
+
+def get_local_path(pkg_uri) -> str:
+    (_, pkg_name) = parse_uri(pkg_uri)
+    return os.path.join(PKG_DIR, pkg_name)
+
+# TODO(yic): Fix this later to handle big directories in better way
+def get_project_package_name(working_dir, modules) -> str:
+    RAY_PKG_PREFIX = "_ray_pkg_"
+    hash_val = None
+    if working_dir:
+        assert isinstance(working_dir, str)
+        assert Path(working_dir).exists()
+        hash_val = _xor_bytes(hash_val, _hash_modules(Path(working_dir)))
+    for module in modules or []:
+        assert inspect.ismodule(module)
+        hash_val = _xor_bytes(hash_val, _hash_modules(Path(module.__file__).parent))
+    return RAY_PKG_PREFIX + hash_val.hex() + ".zip" if hash_val else None
+
+def create_project_package(pkg_file: Path, working_dir: str, modules):
+    with ZipFile(pkg_file, "w") as zip_handler:
+        if working_dir:
+            # put all files in /path/working_dir into zip
+            working_path = Path(working_dir)
+            _zip_module(working_path, working_path, zip_handler)
+        for module in modules or []:
+            logger.info(module.__file__)
+            # we only take care of modules with path like this for now:
+            #    /path/module_name/__init__.py
+            # module_path should be: /path/module_name
+            module_path = Path(module.__file__).parent
+            _zip_module(module_path, module_path.parent, zip_handler)
+
+def parse_uri(pkg_uri: str) -> (Protocol, str):
+    uri = urlparse(pkg_uri)
+    protocol = Protocol(uri.scheme)
+    return (protocol, uri.netloc)
+
+def fetch_package(pkg_uri: str, pkg_file: Path):
+    (protocol, pkg_name) = parse_uri(pkg_uri)
+    if protocol in (Protocol.GCS, Protocol.PIN_GCS):
+        code = internal_kv._internal_kv_get(pkg_uri)
+        code = code or b""
+        pkg_file.write_bytes(code)
+        return len(code)
+    else:
+        raise NotImplementedError(f"Protocol {protocol} is not supported")
+
+def push_package(pkg_uri: str, pkg_file: Path):
```

Comment:
```suggestion
def store_package_in_gcs(pkg_file: Path, delete_with_job: bool) -> str:
```

Reply:
How about this, 

```
def push_package(pkg_uri: str, pkg_file: Path):
    # call the right function to manage it.

def store_package_in_gcs(...):
    # gcs storage implementation

```

I'll do some refactoring when I support S3.

Reply:
Here `pkg_uri` was generated first. It'll be included in job_config.runtime_env first, and then core worker will be constructed.

With core worker constructed, data will be able to be pushed to gcs. So I make `push_pacakge` with uri and file.